### PR TITLE
imap in unified search results

### PR DESCRIPTION
**Pull request by StCyr**
Proof of concept of adding IMAP search to unified search (https://github.com/nextcloud/mail/issues/3481).

Works for INBOX of the user's first account only atm.

What do you think @ChristophWurst ?

### CODE REVIEW

File: lib/Search/Provider.php [hunk 1]
```diff
@@ -76,13 +82,31 @@ public function getOrder(string $route, array $routeParameters): int {
 
 	public function search(IUser $user, ISearchQuery $query): SearchResult {
 		$cursor = $query->getCursor();
+
+		// Search in local cache (doesn't contain the boday of emails)
```
Comment (ChristophWurst) [inline lib/Search/Provider.php:86]:
```suggestion
		// Search in local cache (doesn't contain the body of emails)
```

:wink:

File: lib/Search/Provider.php [hunk 2]
```diff
@@ -76,13 +82,31 @@ public function getOrder(string $route, array $routeParameters): int {
...
 		$messages = $this->mailSearch->findMessagesGlobally(
```
Comment (ChristophWurst) [inline lib/Search/Provider.php:87]:
```suggestion
		$messagesFromCache = $this->mailSearch->findMessagesGlobally(
```

File: lib/Search/Provider.php [hunk 3]
```diff
@@ -76,13 +82,31 @@ public function getOrder(string $route, array $routeParameters): int {
...
 			$user,
 			$query->getTerm(),
 			empty($cursor) ? null : ((int) $cursor),
 			$query->getLimit()
 		);
 
+		// Search in body of emails on the IMAP server
+		$messages2 = $this->mailSearch->findMessagesLocally(
```
Comment (ChristophWurst) [inline lib/Search/Provider.php:95]:
```suggestion
		$messagesFromImap = $this->mailSearch->findMessagesLocally(
```

File: lib/Service/Search/MailSearch.php [hunk 4]
```diff
@@ -136,6 +139,34 @@ public function findMessages(Account $account,
...
+								 ?int $cursor,
+								 ?int $limit): array {
+		$query = $this->filterStringParser->parse($filter);
+		if ($cursor !== null) {
+			$query->setCursor($cursor);
+		}
+
+		$account = $this->accountService->findByUserId($user->getUID())[0];
+
+		$mailbox = $this->mailboxMapper->find($account, 'INBOX');
```
Comment (ChristophWurst) [inline lib/Service/Search/MailSearch.php:163]: this should catch if there is no INBOX (no matter how unlikely that sounds)

**Comment by StCyr** [comment]
HI @ChristophWurst ,

thanks for your review but I think we don't want to include IMAP search in unified search: This will probably slows things to a crawl.

What's your opinion?

**Comment by ChristophWurst** [comment]
That's true. I'm not sure how to best handle it. We could make the body search on IMAP optional. like only if the user searches `body:xyz` but that is not very user friends and almost impossible to discover.

Alternatively there could be a second search provider so the main one does not get slowed down by IMAP search. But that is subpar in terms of UX as there would be two search results for Mail.

**Comment by ChristophWurst** [comment]
I'd say we continue the discussion at https://github.com/nextcloud/mail/issues/3481 until we found an acceptable solution.

I agree that IMAP search for every search query slows down loading of your results significantly for most users. Only those with local mail clients and very fast search implementations will get quick results.

**Pull request closed without merging**